lib/tree.rb

```diff
@@ -2,61 +2,160 @@ class TreeNode
   attr_reader :key, :value
   attr_accessor :left, :right
 
-   def initialize(key, val)
+  def initialize(key, val)
     @key = key
     @value = val
     @left = nil
     @right = nil
-   end
+  end
 end
 
 class Tree
   attr_reader :root
+
   def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
-  def add(key, value)
-    raise NotImplementedError
+  # Time Complexity: O(log(n)) assuming that the tree is balanced, binary search will eliminate half of tree every time, otherwise O(n) like linked list
+  # Space Complexity: same as time complexity for stack calls
+  def add(key, value = nil)
+    @root = add_helper(@root, key, value)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def add_helper(current, key, value)
+    if current.nil?
+      current = TreeNode.new(key, value)
+    elsif key <= current.key
+      current.left = add_helper(current.left, key, value)
+    else
+      current.right = add_helper(current.right, key, value)
+    end
+
+    return current
+  end
+
+  # Time Complexity: O(log(n))/O(n)
+  # Space Complexity: same as time complexity
   def find(key)
-    raise NotImplementedError
+    return find_helper(@root, key)
+  end
+
+  def find_helper(current, key)
+    return nil unless current
+    if key == current.key
+      return current.value
+    elsif key < current.key
+      find_helper(current.left, key)
+    else
+      find_helper(current.right, key)
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity:O(n) we traverse all the nodes once
+  # Space Complexity:O(n) for stack calls + array space(O(2n)) 
   def inorder
-    raise NotImplementedError
+    return inorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def inorder_helper(node, array)
+    return array unless node
+
+    inorder_helper(node.left, array)
+    array << { :key => node.key, :value => node.value }
+    inorder_helper(node.right, array)
+  end
+
+  # Time Complexity:O(n) 
+  # Space Complexity:O(n) 
   def preorder
-    raise NotImplementedError
+    return preorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def preorder_helper(node, array)
+    return array unless node
+
+    array << { :key => node.key, :value => node.value }
+    preorder_helper(node.left, array)
+    preorder_helper(node.right, array)
+  end
+
+  # Time Complexity:O(n) 
+  # Space Complexity:O(n) 
   def postorder
-    raise NotImplementedError
+    return postorder_helper(@root, [])
+  end
+
+  def postorder_helper(node, array)
+    return array unless node
+
+    postorder_helper(node.left, array)
+    postorder_helper(node.right, array)
+    array << { :key => node.key, :value => node.value }
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n) Have to traverse all the nodes once to find the one with the max height
+  # Space Complexity:O(n) for stack calls
   def height
```

Comment:
👍   Your space complexity is O(n) for unbalanced trees and O(log n) for a balanced tree.